Framework/PythonInterface/plugins/algorithms/ANSTO/emu_common.py

```diff
@@ -0,0 +1,254 @@
+        # performs a matrix inversion of the doppler table to create a linear regression
+        try:
+            rows = self._table.split(";")
+            table = []
+            for row in rows:
+                table.append([float(x) for x in row.split(",")])
+            table = np.array(table, dtype=float)
+        except ValueError:
+            raise RuntimeError("Invalid Doppler table entries in file.")
+
+        # the fitting works by finding the the closest entry in the table
+        # that matches the (speed, amp) if either phase or frequency is 'fixed'
+        # if the match is not close raise an exception
+        # note amp is in mm in table
+        speed, amp = self._speed, self._amplitude * 1000.0
+        min_ix = None
+        if self._frequency_option.lower() == "fixed" or self._phase_option.lower() == "fixed":
+            # find the closest match
+            distance = np.linalg.norm(table[:, :2] - [speed, amp], axis=1)
+            min_ix = np.argmin(distance)
+            if distance[min_ix] > 0.01:
+                raise RuntimeError("Cannot find suitable doppler entry for {}, {}.".format(speed, amp))
+
+        if min_ix and self._frequency_option.lower() == "fixed":
+            freq = table[min_ix, 2]
+        elif self._frequency_option.lower() == "auto":
+            freq = 0.5 * speed / (math.pi * self._amplitude)
+        else:
+            freq = float(self._frequency_option)
+
+        if min_ix and self._phase_option.lower() == "fixed":
+            phase = table[min_ix, 3]
+        else:
+            phase = float(self._phase_option)
+
+        self._frequency = freq
+        self._phase = phase
+
+
+class FilterEmuPixelsTubes(FilterPixelsTubes):
+    _doppler_window: str
+
+    def __init__(
+        self,
+        valid_tubes: StrOption,
+        valid_pixels: RangeOption,
+        pixels_per_tube: int,
+        pixel_offset: int,
+        doppler_window: str,
+    ) -> None:
+        FilterPixelsTubes.__init__(self, valid_tubes, valid_pixels, pixels_per_tube, pixel_offset)
+        self._doppler_window = doppler_window
+
+    def set(
+        self,
+        valid_tubes: StrOption = None,
+        valid_pixels: RangeOption = None,
+        pixels_per_tube: IntOption = None,
+        pixel_offset: IntOption = None,
+        doppler_window: StrOption = None,
+    ) -> None:
+        FilterPixelsTubes.set(self, valid_tubes, valid_pixels, pixels_per_tube, pixel_offset)
+        if doppler_window is not None:
+            self._doppler_window = doppler_window
+
+    def _scan_spectra(self, ws_tag: str) -> None:
+        # if the doppler window is not none
+        # for each spectrum get the event list
+        #   - convert the pulse time to doppler time
+        #   - set the mask by the sign of the doppler posn
+        # the include is merged with this code to avoid a repitive
+        # scan over the histograms
+        # TODO get the start time, frequency and doppler pahse for the run
```

Comment:
```suggestion
        # the include is merged with this code to avoid a repetitive
        # scan over the histograms
        # TODO get the start time, frequency and doppler phase for the run
```

Reply:
Ok fixed, 'repitive' understandable if I was a New Zealander.